flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -139,21 +171,40 @@ public AsyncSinkWriter(
         this.maxBatchSize = maxBatchSize;
         this.maxInFlightRequests = maxInFlightRequests;
         this.maxBufferedRequests = maxBufferedRequests;
+        this.flushOnBufferSizeMB = flushOnBufferSizeMB;
+        this.maxTimeInBufferMS = maxTimeInBufferMS;
+
+        this.inFlightRequestsCount = 0;
+        this.bufferedRequestEntriesTotalSizeMB = 0;
+    }
+
+    private void registerCallback() {
+        Sink.ProcessingTimeService.ProcessingTimeCallback ptc =
+                instant -> {
```

Comment:
Which thread is this callback executed in? Is there a chance of race conditions with `existsActiveTimerCallback`?

Reply:
This is always executed in the mailbox thread.

`StreamTask` creates the `ProcessingTimeService` and there is a method that defers the callback to the mailbox, which is provided to `ProcessingTimeService` at construction time.

```
ProcessingTimeCallback deferCallbackToMailbox(
        MailboxExecutor mailboxExecutor, ProcessingTimeCallback callback) {
    return timestamp -> {
        mailboxExecutor.execute(
                () -> invokeProcessingTimeCallback(callback, timestamp),
                "Timer callback for %s @ %d",
                callback,
                timestamp);
    };
}
```